tests/test_config.py

```diff
@@ -492,6 +493,66 @@ def my_function():
         pass
 
 
+def test_config_model_name() -> None:
```

Comment:
Perhaps I'm missing something here, but isn't the only thing that we really need to focus on testing just that the name of the model when assigned via `model_name` is correct? Why even bring in `get_model_name_map`?

You'll definitely want to run a test or two with model names not equivalent to what they would by default be.

Reply:
The goal of this PR was to address https://github.com/pydantic/pydantic/issues/6304, but considering [this](https://github.com/pydantic/pydantic/pull/6814/files/090985bddef2c2d9462ca177ec768f3c77d08cc0#r1451543032) comment, maybe it is better to not touch the `cls.__name__` and change the behaviour of `get_model_name_map` by making it check a new attribute with the `model_name`.

I would rework a bit the implementation in this direction and keep this test. What do you think?